src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java

```diff
@@ -0,0 +1,131 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.ArgumentsProvider;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+import org.junit.platform.commons.util.Preconditions;
+
+/**
+ * Provides a range of {@link Number}s, as defined by an annotation which is its {@link ArgumentsSource}.
+ * Such an annotation should have the following properties:
+ * <ul>
+ *     <li>{@code from} a primitive value for the "start" of the range.</li>
+ *     <li>{@code to} a primitive value for the "end" of the range. {@code to} must have the same type as {@code from}.</li>
+ *     <li>{@code step} a primitive value for the difference between each two values of the range.</li>
+ *     <li>{@code closed} a {@code boolean} value describing if the range includes the last value (cloded), or not (open).</li>
+ * </ul>
+ *
+ * @see IntRangeSource
+ */
+public class RangeSourceProvider implements ArgumentsProvider {
+
+	/**
+	 * {@link Function}s to convert {@link Number}s to primitive wrappers.
+	 */
+	private final List<Number> values = new LinkedList<>();
+
+	/**
+	 * A map from the (wrapper) class to produce, to a function that can produce it from a {@link Number}
+	 */
+	private static final Map<Class<?>, Function<Number, ?>> primitiveMappers = new HashMap<>();
+	static {
+		primitiveMappers.put(Byte.class, Number::byteValue);
+		primitiveMappers.put(Short.class, Number::shortValue);
+		primitiveMappers.put(Integer.class, Number::intValue);
+		primitiveMappers.put(Long.class, Number::longValue);
+		primitiveMappers.put(Float.class, Number::floatValue);
+		primitiveMappers.put(Double.class, Number::doubleValue);
+	}
+
+	@Override
+	public Stream<? extends Arguments> provideArguments(ExtensionContext context) throws Exception {
+		// @formatter:off
+		// since it's a method annotation, the element will always be present
+		List<Annotation> argumentsSources = context
+				.getElement()
+				.map(method -> Arrays
+						.stream(method.getAnnotations())
+						.filter(annotations -> Arrays
+								.stream(annotations.annotationType().getAnnotationsByType(ArgumentsSource.class))
+								.anyMatch(annotation -> getClass().equals(annotation.value())))
+						.collect(Collectors.toList()))
+				.get();
+
+		Preconditions.condition(
+				argumentsSources.size() == 1,
+				() -> String.format(
+						"Expected exactly one annotaiton to provide an ArgumentSource, found %d.",
+						argumentsSources.size()));
+		// @formatter:on
+
+		Annotation argumentsSource = argumentsSources.get(0);
+		Class<? extends Annotation> argumentsSourceClass = argumentsSource.annotationType();
+
+		Method stepMethod = argumentsSourceClass.getMethod("step");
+		Number stepValue = (Number) stepMethod.invoke(argumentsSource);
+		double stepDouble = stepValue.doubleValue();
```

Comment:
This conversation is lossy. `long` contains numbers that `double` can't represent ([apparently](https://stackoverflow.com/a/3793950/2525313) the smallest one is 9,007,199,254,740,993 (2^53 + 1)). Also, `float from + float step` can lead to different results than `(float) ((double) float from + (double) float step)`.

One solution would be to have a `Range` interface with methods that return `Number` and concrete implementations for each primitive type,